server/src/inmantals/server.py

```diff
@@ -57,41 +60,252 @@
 
 
 class InvalidExtensionSetup(Exception):
-    """The extension can only run on a valid project or module, and not on a single file."""
+    """An instance of InmantaLSHandler can only handle one folder (i.e. not a single file) which contains a valid inmanta
+    project or module."""
 
     def __init__(self, message: str) -> None:
         Exception.__init__(self, message)
         self.message = message
 
 
+class Folder:
+    """
+    Wrapper class around a folder (inmanta module or inmanta project). This folder can either be standalone or live in a
+    workspace alongside other folders.
+
+    The constructor should not be called explicitly. Instantiation is done through the unpack_workspaces method
+    :param folder_uri: uri of the folder that is assumed to live in a workspace
+    :param name: name of the folder
+    :param handler: reference to the InmantaLSHandler responsible for this folder
+    """
+
+    folder_uri: str
+    name: str
+    inmanta_project_dir: Union[Optional[tempfile.TemporaryDirectory], str]
+    handler: "InmantaLSHandler"
+
+    def __init__(self, workspace_folder: lsp_types.WorkspaceFolder, handler: "InmantaLSHandler"):
+        folder_uri = urlparse(workspace_folder["uri"])
+
+        self.folder_uri = os.path.abspath(folder_uri.path)
+        self.handler = handler  # Keep a reference to the handler for cleanup
+        self.kind: object
+
+        # Check that we are working inside an existing project:
+        project_file: str = os.path.join(self.folder_uri, module.Project.PROJECT_FILE)
+        if os.path.exists(project_file):
+            self.inmanta_project_dir = os.path.dirname(project_file)
+            self.kind = module.Project
+        else:
+            self.inmanta_project_dir = self.create_tmp_project()
+
+    def get_setting(self, key: str, default: Optional[str] = None) -> Optional[str]:
+        """
+        Get an inmanta extension setting for this folder
+        """
+        if key == "compiler_venv_path":  # support for versions of core that require a compiler venv.
+            return os.path.join(self.folder_uri, ".env-ls-compiler")
+        try:
+            # Look for some custom settings for this folder
+            with open(os.path.join(self.folder_uri, ".vscode", "settings.json")) as settings_file:
+                settings: dict[str, str] = json.load(settings_file)
+                return settings.get(key, default)
+        except FileNotFoundError:
+            return default
+
+    def cleanup(self) -> None:
+        logger.info("Cleaning up %s...", self)
+        if self.inmanta_project_dir:
+            self.inmanta_project_dir.cleanup()
```

Comment:
`self.inmanta_project_dir` could be just a string, no? Is this method no longer meant to be used now that you introduced `register_tmp_project`?

Reply:
> self.inmanta_project_dir could be just a string, no?

Yes that would be much cleaner :+1: 

> Is this method no longer meant to be used now that you introduced register_tmp_project?

Indeed, removing it